packages/web-client/app/models/workflow/workflow-session.ts

```diff
@@ -60,12 +114,98 @@ export default class WorkflowSession {
     );
   }
 
+  getValue<T extends SupportedType>(key: string): T | null {
+    const data: string | null = this._state[key];
+    if (data !== null && data !== undefined) {
+      let json = JSON.parse(data);
+      if (json.type === 'Date') {
+        return new Date(json.value) as T;
+      } else if (json.type === 'BN') {
+        return new BN(json.value) as T;
+      } else {
+        return json.value;
+      }
+    }
+    return null;
+  }
+
+  getValues(): WorkflowSessionDictionary {
+    let result: WorkflowSessionDictionary = {};
+    for (const key in this._state) {
+      const data: string | null = this._state[key];
+      if (data !== null && data !== undefined) {
+        let json = JSON.parse(data);
+        if (json.type === 'Date') {
+          result[key] = new Date(json.value);
+        } else if (json.type === 'BN') {
+          result[key] = new BN(json.value);
+        } else {
+          result[key] = json.value;
+        }
+      }
+    }
+    return result;
+  }
+
+  setValue(key: string, value: SupportedType): void;
+  setValue(hash: Record<string, SupportedType>): void;
+  setValue(
+    hashOrKey: Record<string, SupportedType> | string,
+    value?: SupportedType
+  ): void {
+    if (typeof hashOrKey === 'string') {
+      this.setStateProperty(hashOrKey, value!);
+    } else {
+      for (const key in hashOrKey) {
+        this.setStateProperty(key, hashOrKey[key]);
+      }
+    }
+    // eslint-disable-next-line no-self-assign
+    this._state = this._state; // for reactivity
+    this.persistToStorage();
+  }
+
+  private setStateProperty(key: string, value: SupportedType) {
+    serializeToState(this._state, key, value);
+  }
+
   private persistToStorage(): void {
     if (!this.hasPersistence) return;
 
     this.workflow?.workflowPersistence.persistData(
       this.workflow.workflowPersistenceId,
-      { name: this.workflow.name, state: this.state }
+      { name: this.workflow.name, state: this._state }
     );
   }
 }
+
+export function serializeToState(
+  state: Record<string, string>,
+  key: string,
+  value: SupportedType
+) {
+  if (value instanceof Date) {
+    state[key] = JSON.stringify({
+      value: value.toISOString(),
+      type: 'Date',
+    });
+  } else if (value instanceof BN) {
+    state[key] = JSON.stringify({
+      value: value.toString(),
+      type: 'BN',
+    });
+  } else {
+    state[key] = JSON.stringify({ value });
```

Comment:
Would it be good for this to follow the `{ value, type }` interface too? In case of someone setting a `type` property in the value in workflow session, like `workflowSession.setValue('something', { type: 'Date' })`. We don't have a guard against that.